lib/pbench/test/unit/server/test_datasets_metadata.py

```diff
@@ -401,6 +409,39 @@ def test_put_invalid_deletion(self, query_get_as, query_put_as):
             "user.one": None,
         }
 
+    def test_put_set_errors(self, monkeypatch, query_get_as, query_put_as):
+        real_set = Metadata.setvalue
+
+        @staticmethod
+        def fake_set(dataset: Dataset, key: str, value: Any, user_id: str):
+            if key in ("global.dashboard.seen", "dataset.name"):
+                raise MetadataBadStructure(dataset, key, key)
+            real_set(dataset, key, value, user_id)
```

Comment:
This approach seems a little too "functional".  That is, why does `fake_set()` need to be able to relay to the "real" `set()`?  Why does this test need to call `query_get_as()`?  Can't `fake_set()` do all the verification which is needed without having to actually set anything?

Reply:
All of the API tests are "functional", and I don't see how it makes sense to spend extra time on this particular PR carving out a mock metadata environment for one test case; nor am I going to try to convert the whole test file.

The only thing I might change at this point would be to remove the mock entirely and do it for real as I outlined above. And I'm thinking about it, since you're twisting my arm to do another commit ... it'd be cleaner and more consistent.